Include variable information into assignment, binary and comparison notification events

codingpotato opened the pull request:
- Change `Binary` and `Comparison` events to include variable information.
- Remove `SubscriptAssignment` events, and combine it into `Assignment` events.
- Remove `Boolean` and `Unary` events. It seems that these events are not useful now.

After this change, the `Compare VTag` is not needed any more. Please run following command to see the notification of `Assignment`, `Binary` and `Comparison`.

```shell
dotnet run --project csharp/src/SeedLang.Shell -- -v "*" -f examples/seedpython/scripts/sorting/bubble_sort.py
```

wixette commented on csharp/src/SeedLang/Visualization/EventData.cs:
```diff
@@ -0,0 +1,101 @@
+  // an element of containers or a temporary value. The Variable field is null if it's a temporary
+  // value.
+  public class Operand {
+    public bool IsVariable => !(Variable is null);
```
The structure of `Operand` and `Variable` is a bit confusing:

First, the current `Operand` is classified to:
- Is a Variable
- Is NOT a Variable (that is, a temporary value)

Second, the current `Variable` is classified to:
- Is an Element of a Container
- Is NOT an Element of a Container (that is, a normal variable)

Theoretically "an element of a container" is not always treated as a variable, right?

Can we define an Enum with three categories to classify an operand?
- Variable operand
- ElementOfContainer operand
- TemporaryValue operand

codingpotato replied:
Fixed. Please see my comments seperately.

codingpotato commented on csharp/src/SeedLang/Interpreter/CompilerHelper.cs:
```diff
@@ -195,21 +203,14 @@ internal class CompilerHelper {
                                                   TextRange range) {
-      if (!_suspendNotificationEmitting &&
-          _visualizerCenter.HasVisualizer<Event.SubscriptAssignment>()) {
+      // Check if there are assignment events registered in the visualizer center, because SeedVM
```
Done.

codingpotato commented:
Fixed the review comments and added another big change to fix bugs in variable tracking.

1. Change the event data of visualization notification to LValue and RValue, so that the right value of assignment can also be represented as `Variable`, `Element of Containers` and `Temporary Value`.
2. Added temporary register allocated notification to make variable tracking correct. It will add more instructions when the client would like to get variable related notifications. It makes the program ~5 times slower than normal case without variable tracking enabled. The performance is acceptable now.
3. The variable tracking logic is quite complex. So there are some minor bugs now. I will check and fix them later.
4. Please run following command to see the result of this PR.
```shell
dotnet run --project csharp/src/SeedLang.Shell -- -v "Assignment,Comparison" -f examples/seedpython/scripts/sorting/bubble_sort.py
```

With this PR, the comparison notification events look like
```text
6             if a[j] > a[j + 1]:
Comparison: a:Global[0] 0 > a:Global[1] 1 = False
```
And assignment notification events look like
```text
8                 a[j], a[j + 1] = a[j + 1], a[j]
Assign: a:Global[5] = a:Global[6] 1
8                 a[j], a[j + 1] = a[j + 1], a[j]
Assign: a:Global[6] = a:Global[5] 5
```
Please have a review again.

wixette commented on csharp/src/SeedLang/Visualization/EventData.cs:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  // The LValue class. It could be a variable or an element of containers. The Keys field is null if
+  // it's a variable.
```
For a better interface for this class, it's also good to provide properties to let the client code easily determine the type. E.g.,

```
public bool IsVariable => key is null;
public bool IsElement => !IsVariable;
```

codingpotato replied:
Done.

wixette commented on csharp/src/SeedLang/Visualization/EventData.cs:
```diff
@@ -0,0 +1,144 @@
+
+  // The RValue class. It could be a variable, an element of containers or a temporary value. The
+  // Keys field is null if it's a variable. The Variable and Keys field is null if it's a temporary
+  // value.
```
As above:

```
bool IsVariable => keys is null;
bool IsTemporary => Variable is null && Keys is null;
```

codingpotato replied:
Done.

wixette commented on csharp/src/SeedLang/Visualization/EventData.cs:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  // The RValue class. It could be a variable, an element of containers or a temporary value. The
+  // Keys field is null if it's a variable. The Variable and Keys field is null if it's a temporary
```
```suggestion
  // Keys field is null if it's a variable. The Variable and Keys field are null if it's a temporary
```

codingpotato replied:
Done.

wixette reviewed (approved):
LGTM
